numba/core/event.py

```diff
@@ -0,0 +1,357 @@
+import abc
+import enum
+import time
+from timeit import default_timer as timer
+from contextlib import contextmanager, ExitStack
+from collections import defaultdict
+
+
+class EventStatus(enum.Enum):
+    """Status of an event.
+    """
+    START = enum.auto()
+    END = enum.auto()
+
+
+# Builtin event kinds.
+_builtin_kinds = frozenset([
+    "numba:compiler_lock",
+    "numba:compile",
+])
+
+
+def _guard_kind(kind):
+    """Guard that event kind is valid.
+
+    All event kind with a "numba:" prefix must be defined in pre-defined.
+    Custom event kind is allowed by not using the above prefix.
+
+    Parameters
+    ----------
+    kind : str
+
+    Return
+    ------
+    res : str
+    """
+    if kind.startswith("numba:") and kind not in _builtin_kinds:
+        raise ValueError(f"{kind} is not a valid event kind")
+    return kind
+
+
+class Event:
+    """An event.
+    """
+    def __init__(self, kind, status, data=None, exc_details=None):
+        """
+        Parameters
+        ----------
+        kind : str
+        status : EventStatus
+        data : any; optional
+            Additional data for the event.
+        exc_details : 3-tuple; optional
+            Same 3-tuple for ``__exit__``.
+        """
+        self._kind = _guard_kind(kind)
+        self._status = status
+        self._data = data
+        self._exc_details = exc_details
+
+    @property
+    def kind(self):
+        """Event kind
+
+        Returns
+        -------
+        res : str
+        """
+        return self._kind
+
+    @property
+    def status(self):
+        """Event status
+
+        Returns
+        -------
+        res : EventStatus
+        """
+        return self._status
+
+    @property
+    def data(self):
+        """Event data
+
+        Returns
+        -------
+        res : object
+        """
+        return self._data
+
+    @property
+    def is_start(self):
+        """Is it a *START* event?
+
+        Returns
+        -------
+        res : bool
+        """
+        return self._status == EventStatus.START
+
+    @property
+    def is_end(self):
+        """Is it an *END* event?
+
+        Returns
+        -------
+        res : bool
+        """
+        return self._status == EventStatus.END
+
+    @property
+    def is_failed(self):
+        """Does the event carrying an exception?
+
+        This is used for *END* event. This method will never return ``True``
+        in a *START* event.
+
+        Returns
+        -------
+        res : bool
+        """
+        return self._exc_details[0] is None
+
+    def __str__(self):
+        data = (f"{type(self.data).__qualname__}"
+                if self.data is not None else "None")
+        return f"Event({self._kind}, {self._status}, data: {data})"
+
```

Comment:
It may make sense to add `__repr__` here and have it simply call `__str__`. Also, perhaps a simple `__hash__` implementation could be added that would allow using the `Event` class in dictionaries.

Reply:
Added `__repr__` in https://github.com/numba/numba/pull/6626/commits/1b814b86df9c9b272c32643999a88ba75b66245b.
But `__hash__` will be tricky because `._data` and `._exc_details` may not be hashable. Hashing on just `._kind` and and `._status` is not that helpful. I don't know what behavior it should have in a `dict` or `set`. Furthermore, I think `Event` should only be used in a sequence for replaying the events. Users should even avoid storing events for a long period.

Reply:
sounds good.